cpp/src/arrow/compute/kernels/scalar_cast_numeric.cc

```diff
@@ -769,6 +770,36 @@ std::vector<std::shared_ptr<CastFunction>> GetNumericCasts() {
   return functions;
 }
 
+
+Status CastToExtension(KernelContext* ctx, const ExecSpan& batch, ExecResult* out) {
+  const CastOptions& options = checked_cast<const CastState*>(ctx->state())->options;
+  DCHECK(batch[0].is_array());
+  std::shared_ptr<Array> array = batch[0].array.ToArray();
+  std::shared_ptr<Array> result;
+
+  auto out_ty = GetExtensionType("arrow.py_integer_type")->storage_type();
+  RETURN_NOT_OK(Cast(*array, out_ty, options,
+                     ctx->exec_context())
+                    .Value(&result));
+  
+  ExtensionArray extension(options.to_type.GetSharedPtr(), result);
+  out->value = std::move(extension.data());
+  return Status::OK();
+}
```

Comment:
@pitrou @rok This 'works', but need to hard-code the extension type name here. I think if I can convert the `options.to_type.GetSharedPtr()->storage_id()` into a `TypeHolder/DataType` then I may be in the clear. Would either of you be able to help pointer me to the correct approach? 🙏 
```python
In [2]: from pyarrow.tests.test_extension_type import UuidType, IntegerType

In [3]: import pyarrow as pa

In [4]: pa.register_extension_type(IntegerType())

In [5]: a = pa.array([1, 2, 3], pa.int64())

In [6]: a.cast(IntegerType()).type
Out[6]: IntegerType(DataType(int64))

In [7]: a.cast(IntegerType())
Out[7]:
<pyarrow.lib.ExtensionArray object at 0x7f1167289520>
[
  1,
  2,
  3
]
```

Reply:
I can make it work by templating `GetCastToExtension` and `CastToExtension` then adding the kernel as such:
```cpp
std::vector<std::shared_ptr<CastFunction>> GetExtensionCasts() {
 auto func = GetCastToExtension<Int64Type>("cast_extension"); 
 return {func};                                                                                 
}
```
Where the templated type is the possible storage types of the extension type. Maybe there is a more slick way to accomplish it, especially since the extension name is in the registry, as demonstrated by the hard-coded impl; but will continue with this for now. :)

Reply:
Disregard. Knew as soon as I reached out for help, I'd find a better way. 😆 
Found I could just static cast the `options.to_type.type` to get a  `ExtensionType` (https://github.com/apache/arrow/pull/14106/commits/6c05a1d620dbaea44e2db4520acb16330332df9a)
Apologies for the noise. :)

Reply:
This seems ok and slick-er way can wait.
I'm wondering if this passes enough information about the type. Namely would it work for [parametric extension types](https://github.com/apache/arrow/blob/master/cpp/src/arrow/extension_type_test.cc#L302).

Reply:
Okay, I'll keep that in mind as I clean up the implementation, thanks. 👍

Reply:
I didn't see https://github.com/apache/arrow/pull/14106#discussion_r972979934 when I replied last. I think that approach would probably work.